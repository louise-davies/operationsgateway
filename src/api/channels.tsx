import React from 'react';
import axios, { AxiosError } from 'axios';
import {
  FullChannelMetadata,
  FullScalarChannelMetadata,
  isChannelMetadataScalar,
  timeChannelName,
} from '../app.types';
import {
  useQuery,
  UseQueryResult,
  UseQueryOptions,
} from '@tanstack/react-query';
import { Column } from 'react-table';
import {
  renderImage,
  roundNumber,
} from '../table/cellRenderers/cellContentRenderers';
import { selectUrls } from '../state/slices/configSlice';
import { useAppSelector } from '../state/hooks';
import { readSciGatewayToken } from '../parseTokens';

interface ChannelsEndpoint {
  channels: {
    [systemName: string]: Omit<FullChannelMetadata, 'systemName'>;
  };
}

// This metadata is always present in every record
export const staticChannels: { [systemName: string]: FullChannelMetadata } = {
  [timeChannelName]: {
    systemName: timeChannelName,
    name: 'Time',
    type: 'scalar',
    path: '/system',
  },
  shotnum: {
    systemName: 'shotnum',
    name: 'Shot Number',
    type: 'scalar',
    path: '/system',
  },
  activeArea: {
    systemName: 'activeArea',
    name: 'Active Area',
    type: 'scalar',
    path: '/system',
  },
  activeExperiment: {
    systemName: 'activeExperiment',
    name: 'Active Experiment',
    type: 'scalar',
    path: '/system',
  },
};

const fetchChannels = (apiUrl: string): Promise<FullChannelMetadata[]> => {
  return axios
    .get<ChannelsEndpoint>(`${apiUrl}/channels`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      const { channels } = response.data;

      if (!channels || Object.keys(channels).length === 0) return [];

      const convertedChannels: FullChannelMetadata[] = Object.entries(
        channels
      ).map(([systemName, channel]) => ({
        systemName,
        ...channel,
      }));

      return [...Object.values(staticChannels), ...convertedChannels];
    });
};

export interface ChannelSummary {
  first_date: string;
  most_recent_date: string;
  recent_sample: { [timestamp: string]: string | number }[];
}

const fetchChannelSummary = (
  apiUrl: string,
  channel: string
): Promise<ChannelSummary> => {
  return axios
    .get(`${apiUrl}/channels/summary/${channel}`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useChannels = <T extends unknown = FullChannelMetadata[]>(
  options?: UseQueryOptions<FullChannelMetadata[], AxiosError, T, string[]>
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery(
    ['channels'],
    (params) => {
      return fetchChannels(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      ...(options ?? {}),
    }
  );
};

export const useChannelSummary = (
  channel: string | undefined
): UseQueryResult<ChannelSummary, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const dataChannel =
    typeof channel !== 'undefined' && !(channel in staticChannels)
      ? channel
      : '';

  return useQuery(
    ['channelSummary', dataChannel],
    (params) => {
      return fetchChannelSummary(apiUrl, dataChannel);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      enabled: dataChannel.length !== 0,
    }
  );
};

export const constructColumns = (channels: FullChannelMetadata[]): Column[] => {
  const myColumns: Column[] = [];

  channels.forEach((channel: FullChannelMetadata) => {
    const newColumn: Column = {
      Header: () => {
        const headerName = channel.name ? channel.name : channel.systemName;
        // Provide an actual header here when we have it
        // TODO: do we need to split on things other than underscore?
        const parts = headerName.split('_');
        const wordWrap = parts.map(
          (part, i) =>
            // \u200B renders a zero-width space character
            // which allows line-break but isn't visible
            part + (i < parts.length - 1 ? '_\u200B' : '')
        );
        return <React.Fragment>{wordWrap.join('')}</React.Fragment>;
      },
      accessor: channel.systemName,
      // TODO: get these from data channel info
      channelInfo: channel,
    };
    if (isChannelMetadataScalar(channel)) {
      newColumn.Cell = ({ value }) =>
        typeof value === 'number' && typeof channel.precision === 'number' ? (
          <React.Fragment>
            {roundNumber(value, channel.precision, channel.notation)}
          </React.Fragment>
        ) : (
          <React.Fragment>{String(value ?? '')}</React.Fragment>
        );
    } else {
      newColumn.Cell = ({ row, value }) =>
        renderImage(
          value,
          `${channel.name ?? channel.systemName} ${
            channel.type
          } for timestamp ${row.values[timeChannelName]}`
        );
    }
    myColumns.push(newColumn);
  });
  return myColumns;
};

export const getScalarChannels = (
  channels: FullChannelMetadata[]
): FullScalarChannelMetadata[] => {
  return channels.filter(
    (channel) => channel.type === 'scalar'
  ) as FullScalarChannelMetadata[];
};

const useScalarChannelsOptions = {
  select: (data: FullChannelMetadata[]) => getScalarChannels(data),
};

const useAvailableColumnsOptions = {
  select: (data: FullChannelMetadata[]) => constructColumns(data),
};

export const useScalarChannels = (): UseQueryResult<
  FullScalarChannelMetadata[],
  AxiosError
> => {
  return useChannels(useScalarChannelsOptions);
};

export const useAvailableColumns = (): UseQueryResult<Column[], AxiosError> => {
  return useChannels(useAvailableColumnsOptions);
};
