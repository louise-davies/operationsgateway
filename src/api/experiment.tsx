import { useQuery, UseQueryResult } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { ExperimentParams } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const fetchExperiment = (apiUrl: string): Promise<ExperimentParams[]> => {
  return axios
    .get(`${apiUrl}/experiments`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useExperiment = (): UseQueryResult<
  ExperimentParams[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['experiments'],

    queryFn: () => fetchExperiment(apiUrl),
  });
};
