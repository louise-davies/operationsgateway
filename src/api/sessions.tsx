import axios, { AxiosError } from 'axios';
import { useMutation, UseMutationResult } from '@tanstack/react-query';
import { SaveSessionResponse, Session } from '../app.types';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';
import { readSciGatewayToken } from '../parseTokens';

const saveSession = (
  apiUrl: string,
  session: Session
): Promise<SaveSessionResponse> => {
  const queryParams = new URLSearchParams();
  queryParams.append('name', session.name);
  queryParams.append('summary', session.summary);
  queryParams.append('auto_saved', session.auto_saved.toString());

  return axios
    .post<SaveSessionResponse>(`${apiUrl}/sessions`, session.session_data, {
      params: queryParams,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => response.data);
};

export const useSaveSession = (): UseMutationResult<
  SaveSessionResponse,
  AxiosError,
  Session
> => {
  const { apiUrl } = useAppSelector(selectUrls);
  return useMutation((session: Session) => saveSession(apiUrl, session), {
    onError: (error) => {
      console.log('Got error ' + error.message);
    },
  });
};
