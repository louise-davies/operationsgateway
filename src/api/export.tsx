import axios, { AxiosError } from 'axios';
import { useMutation, UseMutationResult } from '@tanstack/react-query';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';
import { selectSelectedRows } from '../state/slices/selectionSlice';
import { selectQueryParams } from '../state/slices/searchSlice';
import { selectSelectedIdsIgnoreOrder } from '../state/slices/tableSlice';
import { readSciGatewayToken } from '../parseTokens';
import { SearchParams, SortType, timeChannelName } from '../app.types';
import { staticChannels } from './channels';

interface DataToExport {
  Scalars: boolean;
  Images: boolean;
  'Waveform CSVs': boolean;
  'Waveform Images': boolean;
}

export const exportData = (
  apiUrl: string,
  sort: SortType,
  searchParams: SearchParams,
  filters: string[],
  offsetParams?: {
    startIndex: number;
    stopIndex: number;
  },
  projection?: string[],
  dataToExport?: DataToExport,
  selectedRows?: string[]
): Promise<void> => {
  const queryParams = new URLSearchParams();

  for (const [key, value] of Object.entries(sort)) {
    // API recognises sort values as metadata.key or channel.key
    // Therefore, we must construct the appropriate parameter
    const sortKey =
      key in staticChannels ? `metadata.${key}` : `channels.${key}`;
    queryParams.append('order', `${sortKey} ${value}`);
  }

  const { dateRange } = searchParams;

  let timestampObj = {};
  if (dateRange.fromDate || dateRange.toDate) {
    timestampObj = {
      'metadata.timestamp': {
        $gte: dateRange.fromDate,
        $lte: dateRange.toDate,
      },
    };
  }

  const filtersObj = filters
    .filter((f) => f.length !== 0)
    .map((f) => JSON.parse(f));

  const searchObj = [];
  if (dateRange.fromDate || dateRange.toDate) searchObj.push(timestampObj);

  searchObj.push(...filtersObj);

  const existsConditions: { [x: string]: { $exists: boolean } }[] = [];

  projection?.forEach((channel) => {
    // API recognises projection values as metadata.key or channel.key
    // Therefore, we must construct the appropriate parameter
    const key =
      channel in staticChannels ? `metadata.${channel}` : `channels.${channel}`;
    queryParams.append('projection', key);

    if (channel !== timeChannelName) {
      existsConditions.push({ [key]: { $exists: true } });
    }
  });

  if (selectedRows) {
    searchObj.push({ _id: { $in: selectedRows } });
  }

  if (existsConditions.length > 0 || searchObj.length > 0) {
    const query =
      existsConditions.length > 0 && searchObj.length > 0
        ? { $and: searchObj, $or: existsConditions }
        : existsConditions.length > 0
          ? { $or: existsConditions }
          : { $and: searchObj };

    queryParams.append('conditions', JSON.stringify(query));
  }

  if (dataToExport) {
    queryParams.append(
      'export_scalars',
      JSON.stringify(dataToExport['Scalars'])
    );
    queryParams.append('export_images', JSON.stringify(dataToExport['Images']));
    queryParams.append(
      'export_waveform_csvs',
      JSON.stringify(dataToExport['Waveform CSVs'])
    );
    queryParams.append(
      'export_waveform_images',
      JSON.stringify(dataToExport['Waveform Images'])
    );
  }

  if (!(offsetParams?.stopIndex === Infinity)) {
    queryParams.append(
      'skip',
      offsetParams ? JSON.stringify(offsetParams.startIndex) : '0'
    );
    queryParams.append(
      'limit',
      offsetParams
        ? JSON.stringify(offsetParams.stopIndex - offsetParams.startIndex)
        : '0'
    );
  }

  return axios
    .get(`${apiUrl}/export`, {
      params: queryParams,
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
      responseType: 'blob',
    })
    .then((response) => {
      const href = URL.createObjectURL(response.data);
      const link = document.createElement('a');
      link.href = href;
      link.download = response.headers['content-disposition']
        .split('filename=')[1]
        .slice(1, -1);

      link.style.display = 'none';

      document.body.appendChild(link);
      link.click();

      link.remove();
      URL.revokeObjectURL(href);
    });
};

export const useExportData = (): UseMutationResult<void, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const selectedRows = useAppSelector(selectSelectedRows);
  const { searchParams, page, resultsPerPage, sort, filters } =
    useAppSelector(selectQueryParams);
  const projection = useAppSelector(selectSelectedIdsIgnoreOrder);

  const { maxShots } = searchParams;

  return useMutation({
    mutationKey: ['exportData'],

    mutationFn: (params) => {
      const { exportType, dataToExport } = params as {
        exportType: string;
        dataToExport: DataToExport;
      };
      const startIndex =
        exportType === 'Visible Rows' ? page * resultsPerPage : 0;
      const stopIndex =
        exportType === 'Visible Rows'
          ? startIndex + resultsPerPage
          : exportType === 'All Rows'
            ? maxShots
            : 0;

      return exportData(
        apiUrl,
        sort,
        searchParams,
        filters,
        { startIndex, stopIndex },
        projection,
        dataToExport,
        exportType === 'Selected Rows' ? selectedRows : undefined
      );
    },
  });
};
