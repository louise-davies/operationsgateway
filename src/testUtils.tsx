import { Action, ThunkAction } from '@reduxjs/toolkit';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import { matchRequestUrl } from 'msw';
import React from 'react';
// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import { Provider } from 'react-redux';
import { staticChannels } from './api/channels';
import {
  DEFAULT_WINDOW_VARS,
  FullChannelMetadata,
  FullScalarChannelMetadata,
  PlotDataset,
  timeChannelName,
} from './app.types';
import channelsJson from './mocks/channels.json';
import { server } from './mocks/server';
import { COLOUR_ORDER } from './plotting/plotSettings/colourGenerator';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialFilterState } from './state/slices/filterSlice';
import { initialState as initialFunctionsState } from './state/slices/functionsSlice';
import {
  PlotConfig,
  initialState as initialPlotState,
} from './state/slices/plotSlice';
import { initialStateFunc as initialSearchStateFunc } from './state/slices/searchSlice';
import { initialState as initialSelectionState } from './state/slices/selectionSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialWindowsState } from './state/slices/windowSlice';
import { AppStore, RootState, setupStore } from './state/store';
import {
  WindowContext,
  WindowContextProvider,
  WindowsRefType,
} from './windows/windowContext';

/**
 * Waits for msw request -
 * @param method string representing the HTTP method
 * @param url string representing the URL match for the route
 * @returns a promise of the matching request
 *  */
export function waitForRequest(method: string, url: string) {
  let newRequestId = '';

  return new Promise<Request>((resolve, reject) => {
    const onRequestStart = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      const requestURL = new URL(request.url);
      const matchesMethod =
        request.method.toLowerCase() === method.toLowerCase();

      const matchesUrl = matchRequestUrl(requestURL, url).matches;

      if (matchesMethod && matchesUrl) {
        newRequestId = requestId;
      }
    };

    const onRequestMatch = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      if (requestId === newRequestId) {
        server.events.removeListener('request:start', onRequestStart);
        server.events.removeListener('request:match', onRequestMatch);
        server.events.removeListener('request:unhandled', onRequestUnhandled);
        resolve(request);
      }
    };

    const onRequestUnhandled = ({
      request,
      requestId,
    }: {
      request: Request;
      requestId: string;
    }) => {
      const requestURL = new URL(request.url);
      if (requestId === newRequestId) {
        server.events.removeListener('request:start', onRequestStart);
        server.events.removeListener('request:match', onRequestMatch);
        server.events.removeListener('request:unhandled', onRequestUnhandled);
        reject(
          new Error(
            `The ${request.method} ${requestURL.href} request was unhandled.`
          )
        );
      }
    };

    server.events.on('request:start', onRequestStart);

    server.events.on('request:match', onRequestMatch);

    server.events.on('request:unhandled', onRequestUnhandled);
  });
}

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchStateFunc(),
  plots: initialPlotState,
  filter: initialFilterState,
  functions: initialFunctionsState,
  windows: initialWindowsState,
  selection: initialSelectionState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
  queryClient?: QueryClient;
  windowsRef?: React.MutableRefObject<WindowsRefType>;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

export const hooksWrapperWithProviders = (
  state = {},
  queryClient?: QueryClient
) => {
  const testQueryClient = queryClient ?? createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    windowsRef,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    const WindowsProvider = windowsRef
      ? WindowContext.Provider
      : WindowContextProvider;
    return (
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <WindowsProvider {...(windowsRef ? { value: windowsRef } : {})}>
            {children}
          </WindowsProvider>
        </QueryClientProvider>
      </Provider>
    );
  }
  return {
    store,
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderComponentWithStoreAndWindows(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    windowsRef,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    const WindowsProvider = windowsRef
      ? WindowContext.Provider
      : WindowContextProvider;
    if (windowsRef && windowsRef.current === null) windowsRef.current = {};
    return (
      <Provider store={store}>
        <WindowsProvider {...(windowsRef ? { value: windowsRef } : {})}>
          {children}
        </WindowsProvider>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

export const testChannels = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels).map(
    ([systemName, channel]) =>
      ({
        systemName,
        ...channel,
      }) as FullChannelMetadata
  ),
];

export const testScalarChannels: FullScalarChannelMetadata[] = [
  ...Object.values(staticChannels).filter(
    (channel) => channel.type === 'scalar'
  ),
  ...Object.entries(channelsJson.channels)
    .filter(([_systemName, channel]) => channel.type === 'scalar')
    .map(
      ([systemName, channel]) =>
        ({
          systemName,
          ...channel,
        }) as FullScalarChannelMetadata
    ),
];

export const generatePlotDataset = (num: number) => {
  const datasetName = testScalarChannels[num].systemName;
  const plotDataset: PlotDataset = {
    name: datasetName,
    data: [
      {
        timestamp: num,
        [datasetName]: num + num,
      },
      {
        timestamp: num + num,
        [datasetName]: num + num + num,
      },
      {
        timestamp: num + num + num,
        [datasetName]: num + num + num + num,
      },
    ],
  };
  return plotDataset;
};

export const testPlotDatasets = Array.from(Array(3), (_, i) =>
  generatePlotDataset(i + 1)
);

export const generatePlotConfig = (num: number) => {
  const plotTitle = `Plot ${num}`;

  const plotConfig: PlotConfig = {
    id: `test-plot-id-${num}`,
    open: num % 2 === 0,
    title: plotTitle,
    plotType: num % 2 === 0 ? 'scatter' : 'line',
    XAxis: num % 3 === 0 ? timeChannelName : undefined,
    XAxisScale:
      num % 3 === 0 ? 'time' : num % 3 === 1 ? 'linear' : 'logarithmic',
    selectedPlotChannels: [],
    leftYAxisScale: num % 2 === 0 ? 'linear' : 'logarithmic',
    rightYAxisScale: num % 2 === 0 ? 'logarithmic' : 'linear',
    gridVisible: num % 2 === 0,
    axesLabelsVisible: num % 2 !== 0,
    selectedColours: [],
    remainingColours: COLOUR_ORDER.map((colour) => colour),
    ...DEFAULT_WINDOW_VARS,
  };

  return plotConfig;
};

export const testPlotConfigs = Array.from(Array(3), (_, i) =>
  generatePlotConfig(i)
);
