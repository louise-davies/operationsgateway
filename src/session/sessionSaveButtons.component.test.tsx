import {
  type RenderResult,
  act,
  fireEvent,
  screen,
  waitFor,
} from '@testing-library/react';
import { useEditSession, useSaveSession } from '../api/sessions';
import { timeChannelName } from '../app.types';
import { renderComponentWithProviders } from '../testUtils';
import SessionSaveButtons, {
  AUTO_SAVE_INTERVAL_MS,
  SessionsSaveButtonsProps,
} from './sessionSaveButtons.component';

// Mock the useEditSession hook
vi.mock('../api/sessions', () => ({
  useEditSession: vi.fn(),
  useSaveSession: vi.fn(),
}));

describe('session buttons', () => {
  let props: SessionsSaveButtonsProps;
  const onSaveAsSessionClick = vi.fn();
  const onChangeAutoSaveSessionId = vi.fn();
  const createView = (): RenderResult => {
    return renderComponentWithProviders(<SessionSaveButtons {...props} />);
  };

  beforeEach(() => {
    props = {
      onSaveAsSessionClick: onSaveAsSessionClick,
      loadedSessionData: {
        _id: '',
        auto_saved: false,
        name: 'test',
        summary: 'test',
        session: {},
        timestamp: '',
      },
      loadedSessionTimestamp: { timestamp: undefined, autoSaved: undefined },
      onChangeAutoSaveSessionId: onChangeAutoSaveSessionId,
      autoSaveSessionId: undefined,
    };
    vi.useFakeTimers().setSystemTime(new Date('2024-07-15 12:00:00'));

    // Mock the return value of useEditSession hook
    vi.mocked(useEditSession).mockReturnValue({
      mutate: vi.fn().mockResolvedValue({}),
    });
    vi.mocked(useSaveSession).mockReturnValue({
      mutateAsync: vi.fn().mockResolvedValue({}),
    });
  });

  afterEach(() => {
    vi.useRealTimers();
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { asFragment } = createView();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be able to create an autosaved session from the current state of session', () => {
    props = {
      ...props,
      loadedSessionData: {
        name: 'test',
        summary: 'test',
        auto_saved: false,
        session: {},
        _id: '1',
        timestamp: '',
      },
    };
    const { rerender } = createView();

    act(() => {
      vi.advanceTimersByTime(AUTO_SAVE_INTERVAL_MS);
    });

    expect(useSaveSession().mutateAsync).toHaveBeenCalledTimes(1);
    expect(useSaveSession().mutateAsync).toHaveBeenCalledWith({
      auto_saved: true,
      name: 'test (autosaved)',
      summary: 'test',
      session: {
        selection: {
          selectedRows: [],
        },
        table: {
          columnStates: {},
          selectedColumnIds: [timeChannelName],
          page: 0,
          resultsPerPage: 25,
          sort: {},
        },
        search: {
          searchParams: {
            dateRange: {
              fromDate: '2024-07-14T12:00:00',
              toDate: '2024-07-15T12:00:59',
            },
            shotnumRange: {},
            maxShots: 50,
            experimentID: null,
          },
        },
        plots: {},
        filter: { appliedFilters: [[]] },
        windows: {},
      },
    });

    props = {
      ...props,
      loadedSessionData: {
        name: 'test',
        summary: 'test',
        auto_saved: false,
        session: {},
        _id: '2',
        timestamp: '',
      },
    };

    rerender(<SessionSaveButtons {...props} />);

    expect(useSaveSession().mutateAsync).toHaveBeenCalledTimes(1);
    expect(useSaveSession().mutateAsync).toHaveBeenCalledWith({
      auto_saved: true,
      name: 'test (autosaved)',
      summary: 'test',
      session: {
        selection: {
          selectedRows: [],
        },
        table: {
          columnStates: {},
          selectedColumnIds: [timeChannelName],
          page: 0,
          resultsPerPage: 25,
          sort: {},
        },
        search: {
          searchParams: {
            dateRange: {
              fromDate: '2024-07-14T12:00:00',
              toDate: '2024-07-15T12:00:59',
            },
            shotnumRange: {},
            maxShots: 50,
            experimentID: null,
          },
        },
        plots: {},
        filter: { appliedFilters: [[]] },
        windows: {},
      },
    });
  });

  it('should update autosave session when autoSavedSessionId exist', () => {
    props = {
      ...props,
      loadedSessionData: {
        name: 'test',
        summary: 'test',
        auto_saved: false,
        session: {},
        _id: '1',
        timestamp: '',
      },
      autoSaveSessionId: '5',
    };
    createView();

    act(() => {
      vi.advanceTimersByTime(AUTO_SAVE_INTERVAL_MS);
    });

    expect(useEditSession().mutate).toHaveBeenCalledTimes(1);
    expect(useEditSession().mutate).toHaveBeenCalledWith({
      _id: '5',
      auto_saved: true,
      name: 'test (autosaved)',
      summary: 'test',
      session: {
        selection: {
          selectedRows: [],
        },
        table: {
          columnStates: {},
          selectedColumnIds: [timeChannelName],
          page: 0,
          resultsPerPage: 25,
          sort: {},
        },
        search: {
          searchParams: {
            dateRange: {
              fromDate: '2024-07-14T12:00:00',
              toDate: '2024-07-15T12:00:59',
            },
            shotnumRange: {},
            maxShots: 50,
            experimentID: null,
          },
        },
        plots: {},
        filter: { appliedFilters: [[]] },
        windows: {},
      },
      timestamp: '',
    });
  });

  it('should not enable auto save if an user session is not selected', () => {
    createView();

    act(() => {
      vi.advanceTimersByTime(AUTO_SAVE_INTERVAL_MS);
    });

    expect(useEditSession().mutate).not.toHaveBeenCalledTimes(1);
  });

  it('save a user session', async () => {
    props.loadedSessionData = {
      name: 'test',
      summary: 'test',
      auto_saved: false,
      session: {},
      _id: '1',
      timestamp: '',
    };
    createView();
    const saveButton = screen.getByRole('button', { name: 'Save' });
    expect(saveButton).toBeInTheDocument();

    await fireEvent.click(saveButton);

    await waitFor(() => {
      expect(useEditSession().mutate).toHaveBeenCalledTimes(1);
    });
  });

  it('opens the save dialog when there is not a user session selected', async () => {
    props.loadedSessionData = undefined;
    createView();
    const saveAsButton = screen.getByRole('button', { name: 'Save' });
    expect(saveAsButton).toBeInTheDocument();

    await fireEvent.click(saveAsButton);

    await waitFor(() => {
      expect(onSaveAsSessionClick).toHaveBeenCalledTimes(1);
    });
  });
  it('opens the save dialog when save as button is clicked', async () => {
    createView();
    const saveAsButton = screen.getByRole('button', { name: 'Save as' });
    expect(saveAsButton).toBeInTheDocument();

    await fireEvent.click(saveAsButton);

    await waitFor(() => {
      expect(onSaveAsSessionClick).toHaveBeenCalledTimes(1);
    });
  });

  it('shows the last time a selected user session was saved', async () => {
    props = {
      ...props,
      loadedSessionTimestamp: {
        timestamp: '2023-06-29T15:45:00',
        autoSaved: false,
      },
    };
    createView();

    const timestamp = screen.getByTestId('session-save-buttons-timestamp');

    expect(timestamp).toHaveTextContent(
      'Session last saved: 29 Jun 2023 15:45'
    );
  });

  it('shows the last time a selected user session was auto saved', async () => {
    props = {
      ...props,
      loadedSessionTimestamp: {
        timestamp: '2023-06-29T15:45:00',
        autoSaved: true,
      },
    };
    createView();

    const element = screen.getByTestId('session-save-buttons-timestamp');

    expect(element).toHaveTextContent(
      'Session last autosaved: 29 Jun 2023 15:45'
    );
  });
});
