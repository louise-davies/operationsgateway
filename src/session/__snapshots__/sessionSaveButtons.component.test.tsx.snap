// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session buttons renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-mcd6si"
  >
    <div
      class="MuiBox-root css-k008qs"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        data-testid="session-save-buttons-timestamp"
        style="font-size: small; margin: 8px;"
      >
        <span
          style="font-weight: bold;"
        />
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1owl11-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Save
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1owl11-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Save as
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
